akka-actor/src/main/scala/akka/io/TcpOutgoingConnection.scala

```diff
@@ -101,15 +103,23 @@ private[io] class TcpOutgoingConnection(
             } else {
               log.debug("Could not establish connection because finishConnect " +
                 "never returned true (consider increasing akka.io.tcp.finish-connect-retries)")
-              stop()
+              stop(FinishConnectNeverReturnedTrueException)
             }
           }
         }
 
       case ReceiveTimeout ⇒
         if (timeout.isDefined) context.setReceiveTimeout(Duration.Undefined) // Clear the timeout
         log.debug("Connect timeout expired, could not establish connection to [{}]", remoteAddress)
-        stop()
+        stop(connectTimeoutExpired(timeout))
     }
   }
 }
+
+private[io] object TcpOutgoingConnection {
+  val FinishConnectNeverReturnedTrueException =
+    new ConnectException("Could not establish connection because finishConnect never returned true") with NoStackTrace
```

Comment:
At least for this one it should be ok as the `CommandFailed` contains the complete `Complete` command which already contains address information.